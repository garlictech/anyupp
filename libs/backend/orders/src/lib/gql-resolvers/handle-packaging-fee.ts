import { pipe } from 'fp-ts/lib/function';
import * as R from 'ramda';
import { combineLatest, from, Observable } from 'rxjs';
import { concatMap, map, toArray } from 'rxjs/operators';

import { CrudSdk } from '@bgap/crud-gql/api';
import {
  CreateOrderInput,
  OrderItemConfigSet,
  OrderItemInput,
} from '@bgap/domain';
import { throwIfEmptyValue } from '@bgap/shared/utils';

import { calculatePackagingFeeOfOrder } from '../packaging-utils';
import { getUnitProduct } from './utils';

const getNetPackagingFeeOfOrderItem =
  (sdk: CrudSdk) =>
  (item: OrderItemInput): Observable<number> =>
    pipe(
      getUnitProduct(sdk)(item.productId),
      map(genProd => genProd?.variants || []),
      map(R.find(variant => variant?.id === item.variantId)),
      throwIfEmptyValue(`Variant not found: ${item.variantId}`),
      map(variant => variant.netPackagingFee || 0),
    );

const getNetPackagingFeeOfConfigComponent =
  (sdk: CrudSdk) =>
  (
    item: OrderItemInput,
    productSetId: string,
    productComponentId: string,
  ): Observable<number> =>
    pipe(
      getUnitProduct(sdk)(item.productId),
      map(genProd =>
        pipe(
          genProd?.configSets || [],
          R.find(confSet => confSet?.productSetId === productSetId),
          confSet => confSet?.items || [],
          R.find(
            confItem => confItem?.productComponentId === productComponentId,
          ),
        ),
      ),
      throwIfEmptyValue(`Config set not found. productId: ${item.productId}`),
      map(confSet => confSet.netPackagingFee || 0),
    );

const getNetPackagingFeeOfConfigSets =
  (sdk: CrudSdk) =>
  (item: OrderItemInput): Observable<OrderItemConfigSet[] | undefined> =>
    pipe(
      item.configSets || [],
      x => from(x),
      concatMap(configSet =>
        pipe(
          configSet.items,
          x => from(x),
          concatMap(component =>
            getNetPackagingFeeOfConfigComponent(sdk)(
              item,
              configSet.productSetId,
              component.productComponentId,
            ).pipe(
              map(netPackagingFee => ({
                ...component,
                netPackagingFee,
              })),
            ),
          ),
          toArray(),
          map(items => ({
            ...configSet,
            items,
          })),
        ),
      ),
      toArray(),
      map(sets => (R.isEmpty(sets) ? undefined : sets)),
    );

export const addPackagingFeeToOrder =
  (sdk: CrudSdk) =>
  (
    order: CreateOrderInput,
    currency: string,
    taxPercentage?: number | null,
  ): Observable<CreateOrderInput> =>
    pipe(
      order.items,
      x => from(x),
      concatMap(item =>
        pipe(
          combineLatest([
            getNetPackagingFeeOfOrderItem(sdk)(item),
            getNetPackagingFeeOfConfigSets(sdk)(item),
          ]),
          map(([netPackagingFee, configSets]) => ({
            ...item,
            netPackagingFee,
            configSets,
          })),
        ),
      ),
      toArray(),
      map(items => ({
        ...order,
        items,
        packagingSum: {
          netPrice: calculatePackagingFeeOfOrder(items),
          currency: currency,
          taxPercentage: taxPercentage || 0,
        },
      })),
      map(order => ({
        ...order,
        sumPriceShown: {
          ...order.sumPriceShown,
          priceSum:
            order.sumPriceShown.priceSum +
            order.packagingSum.netPrice *
              (1 + order.packagingSum.taxPercentage / 100.0),
          taxSum:
            order.sumPriceShown.taxSum +
            (order.packagingSum.netPrice * order.packagingSum.taxPercentage) /
              100.0,
        },
      })),
    );
