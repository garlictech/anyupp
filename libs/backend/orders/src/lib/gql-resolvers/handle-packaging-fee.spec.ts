/* eslint-disable @typescript-eslint/no-explicit-any */
import { of } from 'rxjs';
import { addPackagingFeeToOrder } from './handle-packaging-fee';

const useCases = [
  {
    label: 'add package fee',
    productId: 'PRODUCTID',
    sdk: {
      GetUnitProduct: jest.fn().mockReturnValue(
        of({
          variants: {
            items: [
              {
                id: 'VARIANTID',
                netPackagingFee: 1,
              },
              {
                id: 'OTHER_VARIANTID',
                netPackagingFee: 2,
              },
              {
                id: 'UNUSED_VARIANTID',
                netPackagingFee: 3,
              },
            ],
          },
          configSets: [
            {
              productSetId: 'PRODUCTSETID1',
              items: [
                {
                  productComponentId: 'PRODUCTCOMPONENTID1',
                  netPackagingFee: 4,
                },
              ],
            },
            {
              productSetId: 'PRODUCTSETID2',
              items: [
                {
                  productComponentId: 'PRODUCTCOMPONENTID2',
                  netPackagingFee: -1,
                },
              ],
            },
          ],
        }),
      ),
    },
  },
  {
    label: 'variant not found',
    productId: 'PRODUCTID2',
    sdk: {
      GetUnitProduct: jest.fn().mockReturnValue(
        of({
          variants: {
            items: [
              {
                id: 'OTHER_VARIANTID',
                netPackagingFee: 2,
              },
              {
                id: 'UNUSED_VARIANTID',
                netPackagingFee: 3,
              },
            ],
          },
          configSets: [
            {
              productSetId: 'PRODUCTSETID1',
              items: [
                {
                  productComponentId: 'PRODUCTCOMPONENTID1',
                  netPackagingFee: 4,
                },
              ],
            },
            {
              productSetId: 'PRODUCTSETID2',
              items: [
                {
                  productComponentId: 'PRODUCTCOMPONENTID2',
                  netPackagingFee: -1,
                },
              ],
            },
          ],
        }),
      ),
    },
  },
  {
    label: 'packaging fee of variant not given',
    productId: 'PRODUCTID3',
    sdk: {
      GetUnitProduct: jest.fn().mockReturnValue(
        of({
          variants: {
            items: [
              {
                id: 'VARIANTID',
              },
              {
                id: 'OTHER_VARIANTID',
              },
            ],
          },
          configSets: [
            {
              productSetId: 'PRODUCTSETID1',
              items: [
                {
                  productComponentId: 'PRODUCTCOMPONENTID1',
                },
              ],
            },
          ],
        }),
      ),
    },
  },
  {
    label: 'config set not found',
    productId: 'PRODUCTID4',
    sdk: {
      GetUnitProduct: jest.fn().mockReturnValue(
        of({
          variants: {
            items: [
              {
                id: 'VARIANTID',
              },
              {
                id: 'OTHER_VARIANTID',
              },
            ],
          },
          configSets: [
            {
              productSetId: 'PRODUCTSETID_OTHER',
              items: [
                {
                  productComponentId: 'PRODUCTCOMPONENTID1',
                },
              ],
            },
          ],
        }),
      ),
    },
  },
  {
    label: 'component not found',
    productId: 'PRODUCTID5',
    sdk: {
      GetUnitProduct: jest.fn().mockReturnValue(
        of({
          variants: {
            items: [
              {
                id: 'VARIANTID',
              },
              {
                id: 'OTHER_VARIANTID',
              },
            ],
          },
          configSets: [
            {
              productSetId: 'PRODUCTSETID1',
              items: [
                {
                  productComponentId: 'OTRHER_PRODUCTCOMPONENTID1',
                },
              ],
            },
          ],
        }),
      ),
    },
  },
  {
    label: 'no config sets',
    productId: 'PRODUCTID6',
    sdk: {
      GetUnitProduct: jest.fn().mockReturnValue(
        of({
          variants: {
            items: [
              {
                id: 'VARIANTID',
              },
              {
                id: 'OTHER_VARIANTID',
              },
            ],
          },
        }),
      ),
    },
  },
];

test.each(useCases)(
  'Adding package fee',
  ({ label, sdk, productId }, done: any) => {
    addPackagingFeeToOrder(sdk as any)(
      {
        id: 'ORDERID',
        items: [
          {
            variantId: 'VARIANTID',
            productId,
            quantity: 1,
            netPackagingFee: -1,
            configSets: [
              {
                productSetId: 'PRODUCTSETID1',
                items: [
                  {
                    productComponentId: 'PRODUCTCOMPONENTID1',
                  },
                ],
              },
            ],
          },
          {
            variantId: 'OTHER_VARIANTID',
            productId: 'OTHER_PRODUCTID',
            quantity: 1,
            netPackagingFee: -2,
          },
        ],
        sumPriceShown: {
          currency: 'HUF',
          priceSum: 100,
          taxSum: 10,
        },
      } as any,
      'HUF',
      10,
    ).subscribe({
      next: res => expect(res).toMatchSnapshot(label),
      error: res => {
        expect(res).toMatchSnapshot(label);
        done();
      },
      complete: () => {
        expect(sdk.GetUnitProduct.mock.calls).toMatchSnapshot(
          `GetUnitProduct calls (${label})`,
        );
        done();
      },
    });
  },
);
