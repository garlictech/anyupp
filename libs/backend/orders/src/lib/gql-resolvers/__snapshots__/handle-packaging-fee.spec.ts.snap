// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Adding package fee: GetUnitProduct calls (add package fee) 1`] = `
Array [
  Array [
    Object {
      "id": "PRODUCTID",
    },
  ],
  Array [
    Object {
      "id": "OTHER_PRODUCTID",
    },
  ],
]
`;

exports[`Adding package fee: GetUnitProduct calls (packaging fee of variant not given) 1`] = `
Array [
  Array [
    Object {
      "id": "PRODUCTID3",
    },
  ],
]
`;

exports[`Adding package fee: add package fee 1`] = `
Object {
  "id": "ORDERID",
  "items": Array [
    Object {
      "configSets": Array [
        Object {
          "items": Array [
            Object {
              "netPackagingFee": 4,
              "productComponentId": "PRODUCTCOMPONENTID1",
            },
          ],
          "productSetId": "PRODUCTSETID1",
        },
      ],
      "netPackagingFee": 1,
      "productId": "PRODUCTID",
      "quantity": 1,
      "variantId": "VARIANTID",
    },
    Object {
      "configSets": undefined,
      "netPackagingFee": 2,
      "productId": "OTHER_PRODUCTID",
      "quantity": 1,
      "variantId": "OTHER_VARIANTID",
    },
  ],
  "packagingSum": Object {
    "currency": "HUF",
    "netPrice": 7,
    "taxPercentage": 10,
  },
  "sumPriceShown": Object {
    "currency": "HUF",
    "priceSum": 107.7,
    "taxSum": 10.7,
  },
}
`;

exports[`Adding package fee: component not found 1`] = `[Error: Config set not found. productId: PRODUCTID5]`;

exports[`Adding package fee: config set not found 1`] = `[Error: Config set not found. productId: PRODUCTID4]`;

exports[`Adding package fee: no config sets 1`] = `[Error: Config set not found. productId: PRODUCTID6]`;

exports[`Adding package fee: packaging fee of variant not given 1`] = `
Object {
  "id": "ORDERID",
  "items": Array [
    Object {
      "configSets": Array [
        Object {
          "items": Array [
            Object {
              "netPackagingFee": 0,
              "productComponentId": "PRODUCTCOMPONENTID1",
            },
          ],
          "productSetId": "PRODUCTSETID1",
        },
      ],
      "netPackagingFee": 0,
      "productId": "PRODUCTID3",
      "quantity": 1,
      "variantId": "VARIANTID",
    },
    Object {
      "configSets": undefined,
      "netPackagingFee": 2,
      "productId": "OTHER_PRODUCTID",
      "quantity": 1,
      "variantId": "OTHER_VARIANTID",
    },
  ],
  "packagingSum": Object {
    "currency": "HUF",
    "netPrice": 2,
    "taxPercentage": 10,
  },
  "sumPriceShown": Object {
    "currency": "HUF",
    "priceSum": 102.2,
    "taxSum": 10.2,
  },
}
`;

exports[`Adding package fee: variant not found 1`] = `[Error: Variant not found: VARIANTID]`;
